File: back-ends/springBoot/src/main/java/com/demo/server/auth/controllers/AuthController.java
package com.demo.server.auth.controllers;

import com.demo.server.auth.dtos.*;
import com.demo.server.auth.exceptions.UserAlreadyExistException;
import com.demo.server.auth.model.User;
import com.demo.server.auth.services.AuthService;
import io.jsonwebtoken.Claims;
import lombok.RequiredArgsConstructor;
import org.jetbrains.annotations.NotNull;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseCookie;
import org.springframework.http.ResponseEntity;
import org.springframework.security.authentication.BadCredentialsException;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/auth")
@RequiredArgsConstructor
public class AuthController {

    private final AuthService authService;

    @GetMapping("/users")
    public List<User> getUsers() {
        return authService.getUsers();
    }

    @GetMapping("/user/{email}")
    public ResponseEntity<?> getUser(@PathVariable @NotNull String email) {
        GetUserDto user = (GetUserDto) authService.getUser(email.toLowerCase(), false);
        if (user == null) {
            SingleMessageDto errMsg = SingleMessageDto.builder()
                    .message("User doesn't exist, incorrect email.")
                    .build();
            return ResponseEntity.status(HttpStatus.NOT_FOUND)
                    .body(errMsg);
        }

        return ResponseEntity.ok(user);
    }

    @PostMapping("/register")
    public ResponseEntity<SingleMessageDto> registerUser(@RequestBody @NotNull RegisterRequestDto req) {
        HttpStatus status;
        String message;

        GetUserDto user = (GetUserDto) authService.getUser(req.getEmail().toLowerCase(), true);
        if (user != null) {
            status = HttpStatus.BAD_REQUEST;
            message = "User already exists." ;
        } else {
            authService.register(req);
            status = HttpStatus.OK;
            message = "User was successfully registered." ;
        }

        SingleMessageDto messageDto = SingleMessageDto.builder()
                .message(message)
                .build();
        return ResponseEntity.status(status)
                .body(messageDto);
    }

    @PostMapping("/login")
    public ResponseEntity<?> loginUser(@RequestBody @NotNull LoginRequestDto req) {
        try {
            NewSessionResponseDto newSessionResponseDto = authService.login(req.getEmail().toLowerCase(), req.getPassword());

            return sendNewSession(newSessionResponseDto);

        } catch (BadCredentialsException e) {
            SingleMessageDto errMsg = SingleMessageDto.builder()
                    .message("Email or password is incorrect.")
                    .build();
            return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(errMsg);
        } catch (UserAlreadyExistException e) {
            SingleMessageDto errMsg = SingleMessageDto.builder()
                    .message(e.getMessage())
                    .build();
            return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(errMsg);
        }
    }

    @PostMapping("/login/google")
    public ResponseEntity<?> loginGoogleUser(@RequestAttribute("googleDecodedClaims") Claims googleDecodedClaims) {
        if (googleDecodedClaims == null) {
            SingleMessageDto errMsg = SingleMessageDto.builder()
                    .message("Google decoded claims is missing!")
                    .build();
            return ResponseEntity.status(HttpStatus.NOT_FOUND)
                    .body(errMsg);
        }

        try {
            NewSessionResponseDto newSessionResponseDto = authService.googleLogin(googleDecodedClaims);

            return sendNewSession(newSessionResponseDto);
        } catch (UserAlreadyExistException e) {
            SingleMessageDto errMsg = SingleMessageDto.builder()
                    .message(e.getMessage())
                    .build();
            return ResponseEntity.status(HttpStatus.CONFLICT)
                    .body(errMsg);
        }

    }

    @PostMapping("/login/github")
    public ResponseEntity<?> loginGithubUser(@RequestBody @NotNull GitHubLoginRequestDto req) {
        try {
            NewSessionResponseDto newSessionResponseDto = authService.githubLogin(req.getCode());

            return sendNewSession(newSessionResponseDto);
        } catch (UserAlreadyExistException e) {
            SingleMessageDto errMsg = SingleMessageDto.builder()
                    .message(e.getMessage())
                    .build();
            return ResponseEntity.status(HttpStatus.CONFLICT)
                    .body(errMsg);
        }
    }


    @GetMapping("/checkSession")
    public ResponseEntity<?> checkSession(@CookieValue("session") String email) {
        // Checks access (session) token from cookie in VerifyAccessTokenFilter...

        try {
            GetUserDto user = authService.check(email);

            LoginResponseDto responseDto = LoginResponseDto.builder()
                    .message("Session status was successfully checked.")
                    .user(user)
                    .build();
            return ResponseEntity.ok(responseDto);

        } catch (UsernameNotFoundException e) {
            SingleMessageDto errMsg = SingleMessageDto.builder()
                    .message("User doesn't exist, incorrect subject within session cookie.")
                    .build();
            return ResponseEntity.status(HttpStatus.NOT_FOUND)
                    .body(errMsg);
        }
    }

    @PostMapping("/refresh")
    public ResponseEntity<?> refreshSession(@CookieValue("refresh") String cookieValue) {
        // When the refresh token is verified generate the new access token.
        try {
            NewSessionResponseDto newSessionResponseDto = authService.refresh(cookieValue);

            return sendNewSession(newSessionResponseDto);
        } catch (UsernameNotFoundException e) {
            SingleMessageDto errMsg = SingleMessageDto.builder()
                    .message("User doesn't exist, incorrect subject in refresh cookie.")
                    .build();
            return ResponseEntity.status(HttpStatus.NOT_FOUND)
                    .body(errMsg);
        }
    }

    private static @NotNull ResponseEntity<LoginResponseDto> sendNewSession(
            @NotNull NewSessionResponseDto dto) {

        LoginResponseDto responseDto = LoginResponseDto.builder()
                .message("User is successfully authenticated.")
                .user(dto.getUser())
                .build();
        return ResponseEntity.ok()
                .header(HttpHeaders.SET_COOKIE,
                        String.valueOf(dto.getAccessCookie()),
                        String.valueOf(dto.getRefreshCookie())
                )
                .body(responseDto);
    }

    @PostMapping("/logout")
    public ResponseEntity<Object> logoutUser() {
        ResponseEntity.BodyBuilder resBuilder = ResponseEntity.ok();
        clearSession(resBuilder);
        SecurityContextHolder.clearContext();

        SingleMessageDto messageDto = SingleMessageDto.builder()
                .message("User session cookies cleared, log out successful.")
                .build();
        return resBuilder.body(messageDto);
    }

    private static void clearSession(ResponseEntity.@NotNull BodyBuilder resBuilder) {
        ResponseCookie accessCookie = ResponseCookie.from("session", "")
                .maxAge(0)
                .httpOnly(true)
                .secure(true)
                .sameSite("None")
                .path("/")
                .build();
        ResponseCookie refreshCookie = ResponseCookie.from("refresh", "")
                .maxAge(0)
                .httpOnly(true)
                .secure(true)
                .sameSite("None")
                .path("/")
                .build();

        resBuilder.header(HttpHeaders.SET_COOKIE, accessCookie.toString(), refreshCookie.toString());
    }

    @DeleteMapping("/user/{email}")
    public ResponseEntity<String> deleteUser(@PathVariable String email) {
        authService.deleteUser(email);
        return ResponseEntity.ok("User " + email + " was successfully deleted.");
    }
}

File: back-ends/express/src/features/authentication/controllers/authController.ts
import { Request, Response, NextFunction } from "express";
import RegisterRequestDto from "../dtos/RegisterRequestDto";

import * as authService from "../services/authService";
import * as jwtService from "../services/jwtService";
import GetUserDto from "../dtos/GetUserDto";

export const getUsers = async (
  req: Request,
  res: Response,
  next: NextFunction
) => {
  try {
    const users = await authService.getUsers();

    return res.status(200).json(users);
  } catch (error: any) {
    error.reason = "failed to send users.";
    next(error);
  }
};

export const getUser = async (
  req: Request,
  res: Response,
  next: NextFunction
) => {
  try {
    const user = await authService.getUser(req.params.email);
    if (!user)
      return res.status(404).send({
        message: "User doesn't exist, incorrect email.",
      });

    return res.status(200).json(user);
  } catch (error: any) {
    error.reason = "failed to send specified user.";
    next(error);
  }
};

export const register = async (
  req: RegisterRequestDto,
  res: Response,
  next: NextFunction
) => {
  try {
    const user = await authService.getUser(req.body.email);
    if (user) return res.status(400).json({ message: "User already exists." });

    await authService.register(req.body);

    return res
      .status(200)
      .json({ message: "User was successfully registered." });
  } catch (error: any) {
    error.reason = "failed to register user.";
    next(error);
  }
};

export const login = async (
  req: Request,
  res: Response,
  next: NextFunction
) => {
  // Checks if user is valid in verifyUserInCache for standard email login...
  // ---or---
  // Checks if the google user is valid in verifyTokens.verifyGoogleIdToken on /login/google request...
  // ---or---
  // Fetches the GitHub user's access token with the code param in getGithubUserAccessToken on /login/github request...

  try {
    req.gitHubAccessToken &&
      console.log("req.gitHubAccessToken", req.gitHubAccessToken);

    const user = (
      req.googleDecodedClaims
        ? await authService.googleLogin(req.googleDecodedClaims)
        : req.gitHubAccessToken
        ? await authService.githubLogin(req.gitHubAccessToken)
        : req.authUser
    ) as GetUserDto & { githubFetchErr?: any };
    if (user.githubFetchErr) {
      // When requesting for the user with the GitHub access token to GitHub's api...
      if (
        typeof user.githubFetchErr === "string" &&
        user.githubFetchErr.includes("not found")
      ) {
        // On no response data.
        return res.status(404).send({
          message: user.githubFetchErr,
        });
      } else {
        // On a bad status code.
        return res.status(403).send({
          message: "GitHub user ID access token is invalid.",
          ERROR: user.githubFetchErr,
        });
      }
    }

    const generateJWT = new jwtService.GenerateJWT(),
      accessToken = generateJWT.accessToken(user.email),
      refreshToken = await generateJWT.refreshToken(user.email);

    res // Create 'session' cookie.
      .cookie("session", accessToken, {
        maxAge: 1000 * 60 * 15, // 15 minutes.
        httpOnly: true,
        secure: true,
        sameSite: "none",
      })
      // Create refresh cookie.
      .cookie("refresh", refreshToken, {
        maxAge: 1000 * 60 * 60 * 24 * 7, // 1 week.
        httpOnly: true,
        secure: true,
        sameSite: "none",
      });
    return res.status(200).json({
      message: "User is successfully authenticated.",
      user: user,
    });
  } catch (error: any) {
    error.reason = "failed to create user session.";
    next(error);
  }
};

export const checkSession = async (
  req: Request,
  res: Response,
  next: NextFunction
) => {
  try {
    // Verifies access token...

    const user = await authService.getUser(req.decodedClaims!.sub!);
    if (!user)
      return res.status(404).send({
        message: "User doesn't exist, incorrect subject in cookie.",
      });

    return res.status(200).json({
      message: "Session status was successfully checked.",
      user: user,
    });
  } catch (error: any) {
    error.reason = "failed to check user session status.";
    next(error);
  }
};

export const refresh = async (
  req: Request,
  res: Response,
  next: NextFunction
) => {
  try {
    // When the refresh token is verified generate the new access token.

    const user = await authService.getUser(req.decodedClaims!.sub!);
    if (!user)
      return res.status(404).send({
        message: "User doesn't exist, incorrect subject in cookie.",
      });

    const generateJWT = new jwtService.GenerateJWT(),
      accessToken = generateJWT.accessToken(user.email),
      refreshToken = generateJWT.refreshToken(user.email);

    res // Create cookies again.
      .cookie("session", accessToken, {
        maxAge: 1000 * 60 * 15, // 15 minutes.
        httpOnly: true,
        secure: true,
        sameSite: "none",
      })
      .cookie("refresh", refreshToken, {
        maxAge: 1000 * 60 * 60 * 24 * 7, // 1 week.
        httpOnly: true,
        secure: true,
        sameSite: "none",
      });
    return res.status(200).json({
      message: "User is successfully authenticated, session refresh complete.",
      user: user,
    });
  } catch (error: any) {
    error.reason = "failed to refresh user session.";
    next(error);
  }
};

export const logout = async (
  req: Request,
  res: Response,
  next: NextFunction
) => {
  try {
    res.clearCookie("session").clearCookie("refresh");
    return res
      .status(200)
      .json({ message: "User session cookies cleared, log out successful." });
  } catch (error: any) {
    error.reason = "failed to clear user session.";
    next(error);
  }
};

// TODO:
export const deleteUser = async (
  req: Request,
  res: Response,
  next: NextFunction
) => {
  try {
    await authService.deleteUser(req.params.email);

    // return res.status(200).json(users);
  } catch (error) {
    next(error);
  }
};
